### ISSUE TITLE
Warp throws ConnectionClosedByPeer if persistent connection is closed  by client in between requests.

### ISSUE THREAD

**Issue by sopvop** [yesodweb/wai#595]
### Problem
If client closes connection between requests, for example if it's connection cache has timed out,
warp-3.2.9 throws `ConnectionClosedByPeer` exception.

There is no way to tell if this happened in OK place or during request transfer, so you can't decide if exception can be discarded or logged as it may indicate problems elsewhere in infrastructure.

HTTP/1.1 [spec](https://tools.ietf.org/html/rfc2616#section-8.1.4) Is not clear on how persistent connections should be timed out by server or client.
The relevant part of spec:
```
8.1.4  Practical Considerations
...
When a client or server wishes to time-out it SHOULD issue a graceful
   close on the transport connection. Clients and servers SHOULD both
   constantly watch for the other side of the transport close, and
   respond to it as appropriate. If a client or server does not detect
   the other side's close promptly it could cause unnecessary resource
   drain on the network.
```
It is usually interpreted that party which wishes to close connection first closes output channel,
and then waits for input channel to be closed by peer. AFAIK curl does that, but I may be wrong here.

In any case, I don't think that connection closed in between requests is an exceptional state for HTTP.

### How to reproduce
```
curl localhost:8080/something 
# get exception
curl localhost:8080/something localhost:8080/something/else 
# get both responses on same connection and 1 exception.
curl -H "Connection: close" localhost:8080/something
# no exception
```

### Expected behaviour
Warp does not throw bogus `ConnectionClosedByPeer` if client closes connections between requests.

**Comment by kazu-yamamoto** [comment]
In HTTP/1.1, `ConnectionClosedByPeer` is used two places: `headerLines` and `readISource`. The former is a normal case and the latter is an error case. So, we need to fix `headerLines`. To fix this, we have two choices:

1. Throwing a different exception type - we should put a catcher in a upper place but we don't have to modify lower functions drastically.
2. Change structures of lower functions, for instance, returning `IO (Maybe somthing)`

@snoyberg Which would you like?

**Comment by snoyberg** [comment]
I'd probably go for changing the functions, since (I think) this is all internal to Warp. Unless there's reason to be worried about performance from the `Maybe` wrapper, but I doubt the overhead will be greater than installing catchers.

**Comment by snoyberg** [comment]
Looks like I did exactly what @kazu-yamamoto mentioned here in my fix for #618, closing.

**Comment by kazu-yamamoto** [comment]
I'm sorry for not taking time to fix this issue.
I concentrate TLS stuff currently.
